### PR TITLE
Adopt AIP-0203: Field Behavior

### LINKED ISSUE #29: Adopt AIP-0203 Field behavior documentation

### PR DESCRIPTION
**Pull request by rambleraptor**
Fixes #29 

Proto changes in https://github.com/aep-dev/aep/pull/13

It's unclear to me if we need an OpenAPI extension, but I'd guess we do?

## 🍱 Types of changes

What types of changes does your code introduce to AEP? _Put an `x` in the boxes
that apply_

- [ ] Enhancement
- [ ] [New proposal](https://aep.dev/1#workflow)
- [X] Migrated from google.aip.dev
- [ ] Chore / Quick Fix

## 📋 Your checklist for this pull request

Please review the [AEP Style and Guidance](https://aep.dev/style-guide) for
contributing to this repository.

### General

- [X] Basic [Guidance](https://aep.dev/style-guide#guidance) is met.
- [X] Ensure that your PR
      [references AEPs](https://aep.dev/style-guide#referencing-aeps)
      correctly.
- [X] [My code has been formatted](https://aep.dev/contributing#formatting)
      (usually `prettier -w .`)

### Additional checklist for a new AEP

- [ ] A new AEP **should** be no more than two pages if printed out.
- [ ] Ensure that the PR is editable by maintainers.
- [ ] Ensure that [File structure](https://aep.dev/style-guide#file-structure)
      guidelines are met.
- [ ] Ensure that
      [Document structure](https://aep.dev/style-guide#document-structure)
      guidelines are met.

💝 Thank you!

### CODE REVIEW

File: aep/general/0203/aep.md.j2 [hunk 1]
```diff
@@ -1,5 +1,278 @@
...
+### Required
+
+The use of `REQUIRED` indicates that the field **must** be present (and set to
+a non-empty value) on the request or resource.
+
+A field **should** only be described as required if _either_:
+
+- It is a field on a resource that a user provides somewhere as input. In this
+  case, the resource is only valid if a "truthy" value is _stored_.
+  - When [creating](./standard-methods) the resource, a value **must** be
```
Comment (toumorokoshi) [inline aep/general/0203/aep.md.j2:151]: nit: I think this type of enumeration should belong in the standard methods - it's a big backwards to do this type of cross-referencing about a specific standard method back to a lower-level AEP around behavior.

File: aep/general/0203/aep.md.j2 [hunk 2]
```diff
@@ -1,5 +1,278 @@
...
+    provided for the field on the create request.
+  - When [updating](./standard-methods) the resource, the user **may** omit the
+    field provided that the field is also absent from the field mask,
+    indicating no change to the field (otherwise it **must** be provided).
+- It is a field on a request message (a message that is an argument to an RPC,
+  with a name usually ending in `Request`). In this case, a value **must** be
+  provided as part of the request, and failure to do so **must** cause an error
+  (usually `INVALID_ARGUMENT`).
+
+We define the term "truthy" above as follows:
```
Comment (toumorokoshi) [inline aep/general/0203/aep.md.j2:161]: I don't think this section makes sense - we definitely consider empty lists, zeros, and empty messages as valid inputs for a required fields. 

How about just removing this section?

File: aep/general/0203/aep.md.j2 [hunk 3]
```diff
@@ -1,5 +1,278 @@
...
+- For messages, any message with at least one "truthy" field.
+
+Fields **should not** be described as required in order to signify:
+
+- A field which will always be present in a response.
+- A field which is conditionally required in some situations.
+- A field on any message (including messages that are resources) which is never
+  used as user input.
+
+**Note:** In most cases, empty values (such as `false` for booleans, `0` for
```
Comment (toumorokoshi) [inline aep/general/0203/aep.md.j2:174]: this note further emphasizes to me that we need to remove this part of the guidance.

Why have a required boolean, if the value for a required boolean must always be true?

proto3 the optional keyword to differentiate between unset and set values.

File: aep/general/0203/aep.md.j2 [hunk 4]
```diff
@@ -1,5 +1,278 @@
...
+
+The use of `UNORDERED_LIST` on a repeated field of a resource indicates that
+the service does not guarantee the order of the items in the list.
+
+A field **should** be described as an unordered list if the service does not
+guarantee that the order of the elements in the list will match the order that
+the user sent, including a situation where the service will sort the list on
+the user's behalf.
+
+A resource with an unordered list **may** return the list in a stable order, or
```
Comment (toumorokoshi) [inline aep/general/0203/aep.md.j2:189]: is this sentence necessary? it seems pretty clear what this annotation guarantees based on the last two paragraphs.

File: aep/general/0203/aep.md.j2 [hunk 5]
```diff
@@ -1,5 +1,278 @@
...
+- Adding `OUTPUT_ONLY` to an existing field previously accepted as input
+- Adding `INPUT_ONLY` to an existing field previously emitted as output
+- Adding `IMMUTABLE` to an existing field previously considered mutable
+- Removing `OUTPUT_ONLY` from an existing field previously ignored as input
+- Removing `IDENTIFIER` from an existing field.
+
+There are some changes that _are_ backwards compatible, which are as follows:
+
+- Adding `OPTIONAL` to an existing field
+- Adding `IDENTIFIER` to an existing `name` field
```
Comment (toumorokoshi) [inline aep/general/0203/aep.md.j2:211]:
```suggestion
- Adding `IDENTIFIER` to an existing `path` field
```

we use path now for AEPs.

File: aep/general/0203/aep.md.j2 [hunk 6]
```diff
@@ -1,5 +1,278 @@
...
+Modifying field behavior after initial authoring can result in
+backwards-incompatible changes in clients. For example, making an optional
+field required results in backwards-incompatible changes in the method
+signature of an RPC or a resource in a Declarative client. See the
+[Backwards compatibility](#backwards-compatibility) section for more detailed
+compatibility guidance.
+
+## History
+
+In 2023-05 field_behavior was made mandatory. Prior to this change, the
```
Comment (toumorokoshi) [inline aep/general/0203/aep.md.j2:263]: we can get rid of this, since AEPs don't have a strong precedent yet?

**Review by toumorokoshi — CHANGES_REQUESTED** [review]
Thanks! A few notes on this one - really on issues in the existing issues.

File: aep/general/0203/aep.md.j2 (hunk 4 above)
Comment (rambleraptor) [reply aep/general/0203/aep.md.j2:189]: Yeah, I can remove this.

File: aep/general/0203/aep.md.j2 (hunk 3 above)
Comment (rambleraptor) [reply aep/general/0203/aep.md.j2:174]: If we're only supporting proto3, I agree with this.

File: aep/general/0203/aep.md.j2 (hunk 1 above)
Comment (rambleraptor) [reply aep/general/0203/aep.md.j2:151]: I agree with this. I'll make an issue for this.

File: aep/general/0203/aep.md.j2 [hunk 7]
```diff
@@ -1,5 +1,278 @@
...
+}
+```
+
+In the case above, if a `title` is specified, the `text` field is required.
+
+## Vocabulary
+
+### Identifier
+
+The use of `IDENTIFIER` indicates that a field within a resource message is
```
[the remark this answers is not in the document]
Comment (rambleraptor) [reply aep/general/0203/aep.md.j2:57]: Yeah, agree on this.

[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0203/aep.md.j2:57]: cool, I see a couple options:

1. wait until Friday to discuss this idea in depth and potentially fix it in this PR.
2. file an issue and we can pick it up later.

Up to you!

File: aep/general/0203/aep.md.j2 (hunk 3 above)
Comment (toumorokoshi) [reply aep/general/0203/aep.md.j2:165]: In proto2 - what value would a boolean that is always set to true serve? Would the idea be that it's a flag that helps others understand that a new version of the schema is used?

File: aep/general/0203/aep.md.j2 (hunk 7 above)
[the remark this answers is not in the document]
Comment (rambleraptor) [reply aep/general/0203/aep.md.j2:57]: I'll raise it in Friday's meeting.

File: aep/general/0203/aep.md.j2 (hunk 3 above)
Comment (rambleraptor) [reply aep/general/0203/aep.md.j2:165]: I'm re-reading this again and you're right. My brain was substituting "must always be true" for "defaults to true"

File: aep/general/0203/aep.md.j2 [hunk 8]
```diff
@@ -1,5 +1,254 @@
...
+
+### Required
+
+The use of `REQUIRED` indicates that the field **must** be present (and set to
+a non-empty value) on the request.
+
+A field **should** only be described as required if _either_:
+
+- It is a field on a resource that a user provides somewhere as input. In this
+  case, the resource is only valid if a "truthy" value is _stored_.
```
Comment (toumorokoshi) [inline aep/general/0203/aep.md.j2:147]:
```suggestion
- It is a field on a resource that a user provides somewhere as input.
```
Removing the "truthy" statement here - the other sub-bullets LGTM.

File: aep/general/0203/aep.md.j2 (hunk 7 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0203/aep.md.j2:57]: filed an issue: https://github.com/aep-dev/aep.dev/issues/181. Let's raise it for discussion, but won't block the adoption of the AEP.

**Review by toumorokoshi — APPROVED** [review]
One last nit, but otherwise LGTM! I'm approving assuming you'll address the one comment, but will hopefully save you the cycle time on approval. 

Thanks!

File: aep/general/0203/aep.md.j2 [hunk 9]
```diff
@@ -1,5 +1,254 @@
...
+```
+
+In the case above, if a `title` is specified, the `text` field is required.
+
+## Vocabulary
+
+### Identifier
+
+The use of `IDENTIFIER` indicates that a field within a resource message is
+used to identify the resource. It **must** be attached to the `name` field and
```
Comment (mkistler) [inline aep/general/0203/aep.md.j2:58]:
```suggestion
used to identify the resource. It **must** be attached to the `path` field and
```

File: aep/general/0203/aep.md.j2 [hunk 10]
```diff
@@ -1,5 +1,254 @@
...
+  field
+- Removing `REQUIRED` from an existing field
+- Removing `INPUT_ONLY` from an existing field previously excluded in responses
+- Removing `IMMUTABLE` from an existing field previously considered immutable
+
+## Rationale
+
+### Identifier field behavior
+
+Resource names, the primary identifiers for any compliant resource, are never
```
Comment (mkistler) [inline aep/general/0203/aep.md.j2:211]:
```suggestion
Resource paths, the primary identifiers for any compliant resource, are never
```

File: aep/general/0203/aep.md.j2 [hunk 11]
```diff
@@ -1,5 +1,254 @@
...
+fully constructed by the user on create. Such fields are typically assigned
+`OUTPUT_ONLY` field behavior. They are, however, also often consumed as the
+primary identifier in scenarios where the resource itself is the primary
+request payload. Such fields could _not_ be considered `OUTPUT_ONLY`.
+Furthermore, in mutation requests, like Standard Update, the resource name as
```
Comment (mkistler) [inline aep/general/0203/aep.md.j2:216]:
```suggestion
Furthermore, in mutation requests, like Standard Update, the resource path as
```

File: aep/general/0203/aep.md.j2 [hunk 12]
```diff
@@ -1,5 +1,254 @@
...
+the primary identifier cannot be changed in place. Such fields are typically
+assigned `IMMUTABLE` field behavior. These conflicting and context-dependent
+field behaviors meant that a new value was necessary to single out and convey
+the behavior of the resource name field.
```
Comment (mkistler) [inline aep/general/0203/aep.md.j2:220]:
```suggestion
the behavior of the resource path field.
```

File: aep/general/0203/aep.md.j2 [hunk 13]
```diff
@@ -1,5 +1,254 @@
...
+In the case above, if a `title` is specified, the `text` field is required.
+
+## Vocabulary
+
+### Identifier
+
+The use of `IDENTIFIER` indicates that a field within a resource message is
+used to identify the resource. It **must** be attached to the `name` field and
+**must not** be attached to any other field (see [fields representing resource
+names]).
```
Comment (mkistler) [inline aep/general/0203/aep.md.j2:60]:
```suggestion
paths]).
```

File: aep/general/0203/aep.md.j2 [hunk 14]
```diff
@@ -1,5 +1,254 @@
...
+behavior when initially authoring of the API.
+
+Modifying field behavior after initial authoring can result in
+backwards-incompatible changes in clients. For example, making an optional
+field required results in backwards-incompatible changes in the method
+signature of an RPC or a resource in a Declarative client. See the
+[Backwards compatibility](#backwards-compatibility) section for more detailed
+compatibility guidance.
+
+[fields representing resource names]:
```
Comment (mkistler) [inline aep/general/0203/aep.md.j2:249]:
```suggestion
[fields representing resource paths]:
```

File: aep/general/0203/aep.md.j2 [hunk 15]
```diff
@@ -1,5 +1,254 @@
...
+  ./resource-paths#fields-representing-resource-names
```
Comment (mkistler) [inline aep/general/0203/aep.md.j2:250]:
```suggestion
  ./resource-paths#fields-representing-resource-paths
```

**Review by mkistler — CHANGES_REQUESTED** [review]
Need to change "name" to "path" in a bunch of places (I think I caught them all).

After that I think we should merge this. I think this AEP will require extensive changes to generalize the guidance and then add OAS-specifics, so we should do that in a separate PR.

File: aep/general/0203/aep.md.j2 [hunk 16]
```diff
@@ -1,5 +1,253 @@
...
+
+Additionally, a field **should** only be described as input only if it is a
+field in a resource message or a field of a message included within a resource
+message. Notably, fields in request messages (a message which only ever acts as
+an argument to an RPC, with a name usually ending in `Request`) **should not**
+be described as input only because this is already implied.
+
+Potential use cases for input only fields (this is not an exhaustive list) are:
+
+- The `ttl` field as described in AIP-214.
```
Comment (mkistler) [inline aep/general/0203/aep.md.j2:102]:
```suggestion
- The `ttl` field as described in AEP-214.
```

File: aep/general/0203/aep.md.j2 [hunk 17]
```diff
@@ -1,5 +1,253 @@
...
+field on a submessage.
+
+### Output only
+
+The use of `OUTPUT_ONLY` indicates that the field is provided in responses, but
+that including the field in a message in a request does nothing (the server
+**must** clear out any value in this field and **must not** throw an error as a
+result of the presence of a value in this field on input). Similarly, services
+**must** ignore the presence of output only fields in update field masks (see:
+AIP-161).
```
Comment (mkistler) [inline aep/general/0203/aep.md.j2:122]:
```suggestion
AEP-161).
```

**Review by mkistler — APPROVED** [review]
Two more minor fixes, but I'll approve as these are straightforward.

**Pull request merged**